modmail/exts/utils/plugin_manager.py

```diff
@@ -0,0 +1,276 @@
+# original source:
+# https://github.com/python-discord/bot/blob/a8869b4d60512b173871c886321b261cbc4acca9/bot/exts/utils/extensions.py  # noqa: E501
+# MIT License 2021 Python Discord
+import functools
+import logging
+import typing as t
+from enum import Enum
+from pathlib import Path
+
+from discord import Colour, Embed
+from discord.ext import commands
+from discord.ext.commands import Context, group
+
+from modmail import plugins
+from modmail.bot import ModmailBot
+from modmail.log import ModmailLogger
+from modmail.utils.cogs import ExtMetadata
+from modmail.utils.plugin_manager import PLUGINS, unqualify
+
+log: ModmailLogger = logging.getLogger(__name__)
+
+BASE_PATH = Path(plugins.__file__).parent
+BASE_PATH_LEN = len(plugins.__name__.split("."))
+
+EXT_METADATA = ExtMetadata(production=True, develop=True, plugin_dev=True)
+
+
+class Action(Enum):
+    """Represents an action to perform on an plugin."""
+
+    # Need to be partial otherwise they are considered to be function definitions.
+    LOAD = functools.partial(ModmailBot.load_extension)
+    UNLOAD = functools.partial(ModmailBot.unload_extension)
+    RELOAD = functools.partial(ModmailBot.reload_extension)
+
+
+class Plugin(commands.Converter):
+    """
+    Fully qualify the name of an plugin and ensure it exists.
+
+    The * and ** values bypass this when used with the reload command.
+    """
+
+    async def convert(self, ctx: Context, argument: str) -> str:
+        """Fully qualify the name of an plugin and ensure it exists."""
+        # Special values to reload all plugins
+        if argument == "*" or argument == "**":
+            return argument
+
+        argument = argument.lower()
+        plugs = []
+        for ext, _nul in PLUGINS:
+            plugs.append(ext)
+
+        if argument in plugs:
+            return argument
+        elif (qualified_arg := f"{plugins.__name__}.{argument}") in PLUGINS:
+            return qualified_arg
+
+        matches = []
+        for ext in plugs:
+            if argument == unqualify(ext):
+                matches.append(ext)
+
+        if len(matches) > 1:
+            matches.sort()
+            names = "\n".join(matches)
+            raise commands.BadArgument(
+                f":x: `{argument}` is an ambiguous plugin name. "
+                f"Please use one of the following fully-qualified names.```\n{names}```"
+            )
+        elif matches:
+            return matches[0]
+        else:
+            raise commands.BadArgument(f":x: Could not find the plugin `{argument}`.")
+
+
+class PluginManager(commands.Cog, name="Plugin Manager"):
+    """Plugin management commands."""
+
+    def __init__(self, bot: ModmailBot):
+        self.bot = bot
+
```

Comment:
Subclass ExtensionManager and just reuse the code, sans formatting stuff, with super()?

Reply:
Importing ExtensionManager from PluginManager will cause a recursive import when utils.extensions looks at exts.utils.plugin_manager